python/tabby-eval/v2-modal/compute_code_completion.py

```diff
@@ -0,0 +1,201 @@
+            "Content-Type": "application/json",
+            "Accept": "application/json",
+            "Authorization": f'Bearer {token}'
+        }
+
+        try:
+            response = httpx.get(url=url, headers=headers, timeout=5)
+            if response.status_code == 200:
+                logging.info("Server details: {}".format(response.json()))
+                return True
+            else:
+                return False
+        except Exception as e:
+            logging.error(f"Failed to check service health: {e}")
+            return False
+
+    while not modal_tabby_ready():
+        time.sleep(5)
+
+    logging.info("Modal tabby server ready!")
+
+
+def monitor_serve_output(process):
+    while True:
+        line = process.stdout.readline()
+        if not line:
+            break
+        logging.info(line.strip())
+
+
+def start_tabby_server(model):
+    logging.info("Starting tabby server for model {model}".format(model=model))
+
+    modal_env = os.environ.copy()
+    modal_env["MODEL_ID"] = model
+    modal_env["EMBEDDING_MODEL_ID"] = EMBEDDING_MODEL_ID
+
+    process = subprocess.Popen(args=["modal", "serve", "app.py"],
+                               stdout=subprocess.PIPE,
+                               stderr=subprocess.STDOUT,
+                               text=True,
+                               env=modal_env)
+
+    threading.Thread(target=monitor_serve_output, args=(process,)).start()
+
+    return process
+
+
+def send_sigint_to_process(process):
+    try:
+        os.kill(process.pid, signal.SIGINT)
+        logging.info("SIGINT signal sent successfully.")
+    except Exception as e:
+        logging.error(f"Failed to send SIGINT signal: {e}")
+
+
+def send_request_with_retry(url, headers, payload, timeout=10, max_retries=10):
+    retries = 0
+    response = None
+
+    while retries < max_retries:
+        try:
+            response = httpx.post(url=url, headers=headers, content=json.dumps(payload), timeout=timeout)
+            if response.status_code == 200:
+                return response
+            else:
+                retries += 1
+                time.sleep(1)
+        except httpx.RequestError as e:
+            logging.error(f"Get code completion failed: {e}")
+            retries += 1
+            time.sleep(1)
+
+    return response
+
+
+def generate_predictions(endpoint: str, token: str, jsonl_file: str, output_prediction_jsonl_file: str):
+    df = pd.read_json(jsonl_file, lines=True)
+    df_flat = json_normalize(df.to_dict(orient="records"))
+
+    headers = {
+        'Content-Type': 'application/json',
+        'Accept': 'application/json',
+        'Authorization': f'Bearer {token}'
+    }
+
+    predictions = []
+    prediction_status = []
+    for index, row in df_flat.iterrows():
+        payload = {
+            "language": row['language'],
+            "segments": {
+                "prefix": row['segments.prefix'],
+                "suffix": row['segments.suffix']
```

Comment:
For each row, please send two request:
1. request with both prefix / suffix.
2. request with prefix, suffix and cross file context.

See https://raw.githubusercontent.com/yan91083/tabby/2ec1bfa6acfb14b5027a053f3433f37bccfbbd79/experimental/eval/data/CodeLlama-13B/csharp/line_completion_rg1_bm25.jsonl for format of `cross_file_context` field. And it can be passed to tabby api through the `debug_options` raw prompt: https://github.com/TabbyML/tabby/blob/91546659aea47237afd513fb6f8491a04febe1b3/crates/tabby/src/services/completion.rs#L83

Reply:
add two types requests to eval code completion.